packages/react/src/components/Dialog/Dialog.tsx

```diff
@@ -142,7 +144,21 @@ class Dialog extends AutoControlledComponent<WithAsProp<DialogProps>, DialogStat
   triggerRef = React.createRef<HTMLElement>()
 
   getInitialAutoControlledState(): DialogState {
-    return { open: false }
+    return {
+      contentId: _.uniqueId('dialog-content-'),
+      headerId: _.uniqueId('dialog-header-'),
+      open: false,
+    }
+  }
+
+  static getAutoControlledStateFromProps(
+    props: DialogProps,
+    state: DialogState,
+  ): Partial<DialogState> {
+    return {
+      contentId: (props.content && props.content['id']) || state.contentId,
+      headerId: (props.header && props.header['id']) || state.headerId,
+    }
```

Comment:
I really like the changes. It would be great if we can extract this logic, as I think it may be reused in other components as well, of course not as part of this PR.

Reply:
Yep, we should inspect other cases with leaked React abstractions